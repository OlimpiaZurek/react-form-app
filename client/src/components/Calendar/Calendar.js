import React, { Component } from "react";
import PropTypes from 'prop-types';
import {
  format,
  startOfMonth,
  isSameMonth,
  isSameDay,
  addMonths,
  subMonths,
} from 'date-fns';
import {
  MONTH_FORMAT,
  DAY_FORMAT,
  WEEK_DAYS
} from '../../common/consts/calendar'
import { getCalendarRows } from '../../common/utils/calendarRows';
import styles from './Calendar.module.scss';

const DAYS = Array.from(Object.values(WEEK_DAYS));

class Calendar extends Component {
  state = {
    currentMonth: new Date(),
  };

  nextMonth = () => {
    this.setState({
      currentMonth: addMonths(this.state.currentMonth, 1)
    });
  };

  prevMonth = () => {
    this.setState({
      currentMonth: subMonths(this.state.currentMonth, 1)
    });
  };

  get header() {
    return (
      <div className={styles.CalendarHeader}>
        <div 
          className={`${styles.Arrow} ${styles.ArrowLeft}`}
          onClick={this.prevMonth} 
        />
        <div className={`${styles.Col} ${styles.ColCenter}`}>
          <span className={styles.CalendarMonth}>
            {format(this.state.currentMonth, MONTH_FORMAT)}
          </span>
        </div>
        <div 
          className={`${styles.Arrow} ${styles.ArrowRight}`}
          onClick={this.nextMonth}
        />
      </div>
    );
  }

  get days() {
    const days = DAYS.map(day => 
      <div 
        className={`${styles.Col} ${styles.ColCenter}`}
        key={day}
      >
        {day}
      </div>
    );
    return <div className={styles.Row}>{days}</div>;
  }

  getDays = (day) => {
    const { selectedDate } = this.props;
    const { currentMonth } = this.state;
    const monthStart = startOfMonth(currentMonth);
    return (
      <div
        className={`${styles.Col} ${
          !isSameMonth(day, monthStart)
            ? styles.Disabled
            : isSameDay(day, selectedDate) ? styles.Selected : ""
        }`}
        key={day}
        onClick={() => this.props.onDateClick(day)}
      >
        <span className={styles.Num}>{format(day, DAY_FORMAT)}</span>
      </div>
    );
  }

  get cells() {
    const { currentMonth } = this.state;
    return getCalendarRows(currentMonth, DAYS).map(cell => 
      <div className={styles.Row} key={cell}>
        {cell.map(day => this.getDays(day))}
      </div>
    );
  }

  render() {
    return (
      <div className={styles.Calendar}>
        {this.header}
        {this.days}
        {this.cells}
      </div>
    );
  }
}

Calendar.propTypes = {
  onDateClick: PropTypes.func,
  selectedDate: PropTypes.instanceOf(Date),
};

Calendar.defaultProps = {
  selectedDate: new Date()
}

export default Calendar;
