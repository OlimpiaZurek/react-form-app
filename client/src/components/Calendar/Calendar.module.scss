@import '../../styles/global.scss';

.Calendar {
    display: block;
    width: 100%;
    background: $neutral-color;
    border: 1px solid $border-color;
  }

.CalendarHeader {
    display: flex;
	align-items: center;
	justify-content: space-between;
}

.Arrow {
    appearance: none;
	user-select: none;
	outline: none !important;
	display: inline-block;
	position: relative;
	cursor: pointer;
	padding: 0;
	border: none;
	border-top: 0.7em solid transparent;
	border-bottom: 0.7em solid transparent;
	transition: all .25s ease-out;
}

.ArrowLeft {
    border-right: 1.2em solid #ccc;
	left: 1.5rem;
	&:hover {
		border-right-color: #06c;
	}
}

.ArrowRight {
    border-left: 1.2em solid #ccc;
	right: 1.5rem;
	&:hover {
		border-left-color: #06c;
	}
}

.CalendarMonth {
    font-weight: 500;
	font-size: 1.2em;
	color: $main-color;
	text-align: center;
	padding: 0.5em 0.25em;
	word-spacing: 5px;
	user-select: none;
}

/* GRID */

.Row {
    margin: 0;
    padding: 0;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    width: 100%;
    height: 50px;
    border: 1px solid $border-color;
  }
  
  .RowMiddle {
    align-items: center;
  }
  
  .Col {
    flex-grow: 1;
    flex-basis: 0;
    max-width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    border-right: 1px solid $border-color;
  }
  
  .ColStart {
    justify-content: flex-start;
    text-align: left;
  }
  
  .ColCenter {
    justify-content: center;
    text-align: center;
  }
  
  .ColEnd {
    justify-content: flex-end;
    text-align: right;
  }

  .Disabled {
    color: $text-color-light;
    pointer-events: none;
  }

  .Selected {
    background-color: $main-color;
    color: $neutral-color;
  }

  .Num {
    font-size: 82.5%;
    line-height: 1;
    text-align: center;;
    font-weight: 700;
  }