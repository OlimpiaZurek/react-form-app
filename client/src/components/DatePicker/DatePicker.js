import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { debounce } from 'lodash';
import TextInput from '../TextInput/TextInput';
import Calendar from '../Calendar/Calendar';
import { 
  DATE_PICKER_PLACEHOLDER
} from '../../common/consts/calendar';

class DatePicker extends Component {

  constructor(props) {
    super(props);
      this.state = {
        isCalendarOpened: false,
    };
    this.delayedFunc = debounce(this.toggleCalendar, 300)
  }
 
  toggleCalendar = () => {
    this.setState({
      isCalendarOpened: !this.state.isCalendarOpened
    });
  }
  
  handleChange = (e) => {
    e.preventDefault();
  }

  onClick = (day) => {
    this.delayedFunc()
    this.props.onDateClick(day)
  }

  render() {
    const { isCalendarOpened } = this.state;
    const { value } = this.props;
    return (
      <Fragment>
        <TextInput
          type="text"
          value={value}
          placeholder={DATE_PICKER_PLACEHOLDER}
          handleChange={this.handleChange}
          onClick={this.toggleCalendar}
          {...this.props}
        />
        {isCalendarOpened && (
          <Calendar 
            {...this.props}
            onDateClick={this.onClick}
          />
        )}
      </Fragment>
     
    );
  }
}

DatePicker.propTypes = {
 value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Date)
 ]),
 onDateClick: PropTypes.func,
};

export default DatePicker;
