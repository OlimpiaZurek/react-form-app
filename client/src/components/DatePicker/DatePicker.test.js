import React from 'react';
import { shallow, mount } from 'enzyme';
import DatePicker from './Datepicker.js';
import Calendar from '../Calendar/Calendar';

describe('<DatePicker />', () => {
  const mockFn = jest.fn();
  it('should be defined', () => {
      expect(DatePicker).toBeDefined();
    });
  
  it('should render corectly', () => {
    const wrapper = shallow(
      <DatePicker value='2.11.2019' onDateClick={mockFn} />
    );
    expect(wrapper).toMatchSnapshot();
  }); 

  it('check month and years dropdowns displayed', () => {
    const wrapper = shallow(
      <DatePicker value='2.11.2019' onDateClick={mockFn} />
    );
    wrapper.setState({ isCalendarOpened: true})
    expect(wrapper.state().isCalendarOpened).toEqual(true)
  });

  it('should render Calendar component', () => {
    const wrapper = mount(
      <DatePicker value='test' onDateClick={mockFn} />
    );
    expect(wrapper.find('TextInput')).toHaveLength(1)
  });

  it('should not call mock function when calendar is not open', () => {
    const wrapper = mount(
      <DatePicker value='test' onDateClick={mockFn} />
    );
    wrapper.simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(0)
  });
});