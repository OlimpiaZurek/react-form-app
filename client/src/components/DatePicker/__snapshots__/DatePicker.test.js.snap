// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> should render corectly 1`] = `ShallowWrapper {}`;
